File: snapkit-v2/snapkit/temporal.py
"""
Temporal snap — T-minus-0 detection and beat grid alignment.

Detects when a temporal signal reaches a critical point (T-0) and
snaps observations to the nearest beat grid. This is the temporal
analog of the Eisenstein spatial snap: instead of snapping to a
hexagonal lattice in the complex plane, we snap to a 1D grid on
the time axis.

Key concepts:
  - BeatGrid: a periodic grid of time points with configurable phase.
  - T-minus-0: the moment a signal transitions from "approaching" to
    "arrived" — analogous to the snap boundary in spatial algorithms.
  - TemporalSnap: combines beat grid alignment with T-0 detection.
"""

import math
from dataclasses import dataclass, field
from typing import List, Optional, Tuple


@dataclass(frozen=True, slots=True)
class TemporalResult:
    """Result of a temporal snap operation."""
    original_time: float
    snapped_time: float
    offset: float           # signed distance to nearest grid point
    is_on_beat: bool        # |offset| ≤ tolerance
    is_t_minus_0: bool      # T-0 event detected
    beat_index: int         # index of nearest beat (0-based)
    beat_phase: float       # phase within current beat cycle [0, 1)


@dataclass
class BeatGrid:
    """A periodic grid of time points.

    Attributes:
        period: interval between beats (seconds).
        phase: offset of beat 0 from t=0 (seconds, 0 ≤ phase < period).
        t_start: start time of the grid (seconds).
    """
    period: float = 1.0
    phase: float = 0.0
    t_start: float = 0.0

    def nearest_beat(self, t: float) -> Tuple[float, int]:
        """Find the nearest grid point to time t.

        Returns (beat_time, beat_index).
        """
        if self.period <= 0:
            raise ValueError("period must be positive")
        adjusted = t - self.t_start - self.phase
        index_float = adjusted / self.period
        index = round(index_float)
        beat_time = self.t_start + self.phase + index * self.period
        return beat_time, int(index)

    def snap(self, t: float, tolerance: float = 0.1) -> TemporalResult:
        """Snap time t to the nearest beat within tolerance.

        Args:
            t: Time to snap.
            tolerance: Maximum offset for on-beat classification.

        Returns:
            TemporalResult with snap details.
        """
        beat_time, beat_index = self.nearest_beat(t)
        offset = t - beat_time
        is_on_beat = abs(offset) <= tolerance
        phase = ((t - self.t_start - self.phase) % self.period) / self.period
        if phase < 0:
            phase += 1.0

        return TemporalResult(
            original_time=t,
            snapped_time=beat_time,
            offset=offset,
            is_on_beat=is_on_beat,
            is_t_minus_0=False,
            beat_index=beat_index,
            beat_phase=phase,
        )

    def beats_in_range(self, t_start: float, t_end: float) -> List[float]:
        """List all beat times in [t_start, t_end]."""
        if t_end <= t_start:
            return []
        first_idx = math.ceil((t_start - self.t_start - self.phase) / self.period)
        last_idx = math.floor((t_end - self.t_start - self.phase) / self.period)
        return [
            self.t_start + self.phase + i * self.period
            for i in range(first_idx, last_idx + 1)
        ]


class TemporalSnap:
    """Temporal snap with T-minus-0 detection.

    T-0 detection works by monitoring a stream of observations and
    detecting when the signal crosses a threshold (approaches zero
    derivative, or crosses a critical value). This is the temporal
    equivalent of the Eisenstein snap boundary.

    The detector uses a simple sliding-window approach:
    - Track the last N observations.
    - T-0 fires when the derivative crosses zero (local extremum)
      AND the value is within a critical zone.
    """

    def __init__(
        self,
        grid: BeatGrid,
        tolerance: float = 0.1,
        t0_threshold: float = 0.05,
        t0_window: int = 3,
    ):
        """
        Args:
            grid: Beat grid for snap alignment.
            tolerance: Maximum offset for on-beat classification.
            t0_threshold: Value threshold for T-0 zone.
            t0_window: Window size for derivative estimation.
        """
        self.grid = grid
        self.tolerance = tolerance
        self.t0_threshold = t0_threshold
        self.t0_window = max(2, t0_window)
        self._history: List[Tuple[float, float]] = []  # (time, value)

    def observe(self, t: float, value: float) -> TemporalResult:
        """Observe a (time, value) pair and snap.

        T-0 is detected when:
          1. The value is within t0_threshold of zero (critical zone).
          2. The derivative (estimated from window) crosses zero.
        """
        self._history.append((t, value))
        if len(self._history) > self.t0_window * 2:
            self._history = self._history[-self.t0_window * 2:]

        # T-0 detection
        is_t0 = self._detect_t0()

        result = self.grid.snap(t, self.tolerance)
        # Override the T-0 flag from the grid snap
        return TemporalResult(
            original_time=result.original_time,
            snapped_time=result.snapped_time,
            offset=result.offset,
            is_on_beat=result.is_on_beat,
            is_t_minus_0=is_t0,
            beat_index=result.beat_index,
            beat_phase=result.beat_phase,
        )

    def _detect_t0(self) -> bool:
        """Detect T-minus-0 from observation history.

        T-0 fires when:
          - Value is in the critical zone (|value| ≤ threshold).
          - Derivative sign change detected in the window.
        """
        if len(self._history) < 2:
            return False

        # Check if current value is in critical zone
        _, current_val = self._history[-1]
        if abs(current_val) > self.t0_threshold:
            return False

        # Estimate derivative from recent observations
        if len(self._history) >= self.t0_window:
            recent = self._history[-self.t0_window:]
            dt = recent[-1][0] - recent[0][0]
            if dt == 0:
                return False
            dv = recent[-1][1] - recent[0][1]
            deriv_current = dv / dt
        else:
            _, prev_val = self._history[-2]
            _, curr_val = self._history[-1]
            prev_t, curr_t = self._history[-2][0], self._history[-1][0]
            dt = curr_t - prev_t
            if dt == 0:
                return False
            deriv_current = (curr_val - prev_val) / dt

        # Check for derivative sign change (zero crossing)
        if len(self._history) >= 3:
            prev_t, prev_val = self._history[-3]
            mid_t, mid_val = self._history[-2]
            curr_t, curr_val = self._history[-1]
            d1 = (mid_val - prev_val) / (mid_t - prev_t) if (mid_t - prev_t) != 0 else 0
            d2 = (curr_val - mid_val) / (curr_t - mid_t) if (curr_t - mid_t) != 0 else 0
            # Sign change in derivative
            return d1 * d2 < 0

        return False

    def reset(self) -> None:
        """Clear observation history."""
        self._history.clear()

    @property
    def history(self) -> List[Tuple[float, float]]:
        """Read-only access to observation history."""
        return list(self._history)

File: snapkit-zig/src/temporal.zig
//! Temporal snap — BeatGrid, TemporalSnap, T-minus-0 detection.
//!
//! No hidden allocations. BeatGrid is a plain struct. TemporalSnap uses
//! a fixed-size circular buffer (no heap allocation).

const std = @import("std");
const types = @import("types.zig");
const BeatGrid = types.BeatGrid;
const TemporalResult = types.TemporalResult;

pub const T0_MAX_HISTORY: usize = 64;

/// Initialize a BeatGrid. Returns error.InvalidPeriod if period ≤ 0.
pub fn beatGridInit(period: f64, phase: f64, t_start: f64) error{InvalidPeriod}!BeatGrid {
    if (period <= 0.0) return error.InvalidPeriod;
    return .{
        .period = period,
        .phase = phase,
        .t_start = t_start,
        .inv_period = 1.0 / period,
    };
}

/// Find the nearest beat time and its index.
pub fn beatGridNearest(grid: *const BeatGrid, t: f64) struct { time: f64, index: i64 } {
    const adjusted = t - grid.t_start - grid.phase;
    const idx: i64 = @intFromFloat(@round(adjusted * grid.inv_period));
    const beat_time = grid.t_start + grid.phase + @as(f64, @floatFromInt(idx)) * grid.period;
    return .{ .time = beat_time, .index = idx };
}

/// Snap a timestamp to the beat grid.
pub fn beatGridSnap(grid: *const BeatGrid, t: f64, tolerance: f64) TemporalResult {
    const nearest = beatGridNearest(grid, t);
    const offset = t - nearest.time;
    var phase = @mod((t - grid.t_start - grid.phase), grid.period) * grid.inv_period;
    if (phase < 0.0) phase += 1.0;

    return .{
        .original_time = t,
        .snapped_time = nearest.time,
        .offset = offset,
        .is_on_beat = @abs(offset) <= tolerance,
        .is_t_minus_0 = false,
        .beat_index = nearest.index,
        .beat_phase = phase,
    };
}

/// Batch snap timestamps to beat grid. Caller provides output slice.
pub fn beatGridSnapBatch(
    grid: *const BeatGrid,
    timestamps: []const f64,
    tolerance: f64,
    out: []TemporalResult,
) void {
    const len = @min(timestamps.len, out.len);
    for (timestamps[0..len], out[0..len]) |t, *o| {
        o.* = beatGridSnap(grid, t, tolerance);
    }
}

/// Enumerate beats in a time range. Caller provides output buffer.
/// Returns the number of beats written.
pub fn beatGridRange(
    grid: *const BeatGrid,
    t_start: f64,
    t_end: f64,
    out: []f64,
) usize {
    if (t_end <= t_start) return 0;
    const first_idx: i64 = @intFromFloat(@ceil((t_start - grid.t_start - grid.phase) * grid.inv_period));
    const last_idx: i64 = @intFromFloat(@floor((t_end - grid.t_start - grid.phase) * grid.inv_period));
    var count: usize = 0;
    var i: i64 = first_idx;
    while (i <= last_idx and count < out.len) : (i += 1) {
        out[count] = grid.t_start + grid.phase + @as(f64, @floatFromInt(i)) * grid.period;
        count += 1;
    }
    return count;
}

// ── T-minus-0 detection ──

/// Temporal snap with T-minus-0 detection via circular buffer.
/// No heap allocation — fixed-size buffer.
pub const TemporalSnap = struct {
    grid: BeatGrid,
    tolerance: f64,
    t0_threshold: f64,
    t0_window: usize,
    hist_t: [T0_MAX_HISTORY]f64,
    hist_v: [T0_MAX_HISTORY]f64,
    hist_idx: usize,
    hist_len: usize,
    hist_cap: usize,

    pub fn init(
        grid: BeatGrid,
        tolerance: f64,
        t0_threshold: f64,
        t0_window: usize,
    ) @This() {
        const win = @max(2, t0_window);
        var cap = win * 2;
        if (cap > T0_MAX_HISTORY) cap = T0_MAX_HISTORY;
        return .{
            .grid = grid,
            .tolerance = tolerance,
            .t0_threshold = t0_threshold,
            .t0_window = win,
            .hist_t = undefined,
            .hist_v = undefined,
            .hist_idx = 0,
            .hist_len = 0,
            .hist_cap = cap,
        };
    }

    /// Observe a (time, value) pair. Returns TemporalResult with T-minus-0 detection.
    pub fn observe(self: *@This(), t: f64, value: f64) TemporalResult {
        self.hist_t[self.hist_idx] = t;
        self.hist_v[self.hist_idx] = value;
        self.hist_idx = (self.hist_idx + 1) % self.hist_cap;
        if (self.hist_len < self.hist_cap) self.hist_len += 1;

        const is_t0 = self.detectT0();
        var result = beatGridSnap(&self.grid, t, self.tolerance);
        result.is_t_minus_0 = is_t0;
        return result;
    }

    fn detectT0(self: *@This()) bool {
        if (self.hist_len < 3) return false;
        const cap = self.hist_cap;
        const idx = self.hist_idx;

        const curr_t = self.hist_t[(idx - 1 + cap) % cap];
        const curr_v = self.hist_v[(idx - 1 + cap) % cap];
        const mid_t = self.hist_t[(idx - 2 + cap) % cap];
        const mid_v = self.hist_v[(idx - 2 + cap) % cap];
        const prev_t = self.hist_t[(idx - 3 + cap) % cap];
        const prev_v = self.hist_v[(idx - 3 + cap) % cap];

        if (@abs(curr_v) > self.t0_threshold) return false;

        const dt1 = mid_t - prev_t;
        const dt2 = curr_t - mid_t;
        if (dt1 == 0.0 or dt2 == 0.0) return false;

        const d1 = (mid_v - prev_v) / dt1;
        const d2 = (curr_v - mid_v) / dt2;

        return d1 * d2 < 0.0;
    }

    pub fn reset(self: *@This()) void {
        self.hist_idx = 0;
        self.hist_len = 0;
    }
};

// ── C-compatible exports ──

export fn sk_beat_grid_init(grid: *BeatGrid, period: f64, phase: f64, t_start: f64) c_int {
    const g = beatGridInit(period, phase, t_start) catch return -1;
    grid.* = g;
    return 0;
}

export fn sk_beat_grid_nearest(grid: *const BeatGrid, t: f64, beat_index: ?*i64) f64 {
    const nearest = beatGridNearest(grid, t);
    if (beat_index) |bi| bi.* = nearest.index;
    return nearest.time;
}

export fn sk_beat_grid_snap(grid: *const BeatGrid, t: f64, tolerance: f64) TemporalResult {
    return beatGridSnap(grid, t, tolerance);
}

export fn sk_beat_grid_snap_batch(
    grid: *const BeatGrid,
    timestamps: [*]const f64,
    n: usize,
    tolerance: f64,
    out: [*]TemporalResult,
) void {
    beatGridSnapBatch(grid, timestamps[0..n], tolerance, out[0..n]);
}

// ── Tests ──

test "BeatGrid init — valid" {
    const grid = try beatGridInit(1.0, 0.0, 0.0);
    try std.testing.expectEqual(@as(f64, 1.0), grid.period);
    try std.testing.expectEqual(@as(f64, 1.0), grid.inv_period);
}

test "BeatGrid init — invalid period" {
    _ = beatGridInit(-1.0, 0.0, 0.0) catch |err| {
        try std.testing.expectEqual(error.InvalidPeriod, err);
        return;
    };
    unreachable;
}

test "BeatGrid snap — on beat" {
    const grid = try beatGridInit(1.0, 0.0, 0.0);
    const result = beatGridSnap(&grid, 2.0, 0.1);
    try std.testing.expect(result.is_on_beat);
    try std.testing.expectApproxEqAbs(@as(f64, 2.0), result.snapped_time, 1e-10);
    try std.testing.expectApproxEqAbs(@as(f64, 0.0), result.offset, 1e-10);
}

test "BeatGrid snap — off beat" {
    const grid = try beatGridInit(1.0, 0.0, 0.0);
    const result = beatGridSnap(&grid, 2.5, 0.1);
    try std.testing.expect(!result.is_on_beat);
}

test "BeatGrid snap — near beat within tolerance" {
    const grid = try beatGridInit(1.0, 0.0, 0.0);
    const result = beatGridSnap(&grid, 2.05, 0.1);
    try std.testing.expect(result.is_on_beat);
    try std.testing.expectApproxEqAbs(@as(f64, 2.0), result.snapped_time, 1e-10);
}

test "BeatGrid phase" {
    const grid = try beatGridInit(1.0, 0.0, 0.0);
    const result = beatGridSnap(&grid, 2.5, 0.1);
    try std.testing.expectApproxEqAbs(@as(f64, 0.5), result.beat_phase, 1e-10);
}

test "BeatGrid range" {
    const grid = try beatGridInit(1.0, 0.0, 0.0);
    var buf: [10]f64 = undefined;
    const count = beatGridRange(&grid, 1.5, 4.5, &buf);
    try std.testing.expectEqual(@as(usize, 3), count);
    try std.testing.expectApproxEqAbs(@as(f64, 2.0), buf[0], 1e-10);
    try std.testing.expectApproxEqAbs(@as(f64, 3.0), buf[1], 1e-10);
    try std.testing.expectApproxEqAbs(@as(f64, 4.0), buf[2], 1e-10);
}

test "TemporalSnap — T-minus-0 detection" {
    const grid = try beatGridInit(1.0, 0.0, 0.0);
    var ts = TemporalSnap.init(grid, 0.1, 0.05, 3);

    // Rising values
    _ = ts.observe(0.0, 0.1);
    _ = ts.observe(1.0, 0.2);
    // Peak
    _ = ts.observe(2.0, 0.01); // small value near zero

    // After sign change through zero, should detect T-minus-0
    // (This depends on the derivative sign change check)
}

test "TemporalSnap reset" {
    const grid = try beatGridInit(1.0, 0.0, 0.0);
    var ts = TemporalSnap.init(grid, 0.1, 0.05, 3);
    _ = ts.observe(0.0, 0.1);
    _ = ts.observe(1.0, 0.2);
    try std.testing.expectEqual(@as(usize, 2), ts.hist_len);
    ts.reset();
    try std.testing.expectEqual(@as(usize, 0), ts.hist_len);
}

test "TemporalSnap batch" {
    const grid = try beatGridInit(1.0, 0.0, 0.0);
    const timestamps = [_]f64{ 0.0, 1.0, 2.0, 3.0 };
    var results: [4]TemporalResult = undefined;
    beatGridSnapBatch(&grid, &timestamps, 0.1, &results);
    for (&results) |r| {
        try std.testing.expect(r.is_on_beat);
    }
}
